src/GAP/gap_to_oscar.jl

```diff
@@ -25,27 +25,115 @@ end
+        # results in slightly better code
+        val = GAPWrap.INT_FFE_DEFAULT(x)::Int
+        return F(val)
+    end
+
+    error("TODO: add support for FFE outside the prime field")
+end
+
+# test code for producing a gap finite field element not stored as FFE:  `GAP.Globals.Z(65537)`
+function (F::FinField)(x::GapObj)
+    GAP.GAP_IS_INT(x) && return F(fmpz(x))
+    GAPWrap.IsFFE(x) || error("<x> must be a GAP large integer or a GAP finite field element")
+    characteristic(x) == characteristic(F) || error("characteristic does not match")
+
+    if GAPWrap.DegreeFFE(x) == 1
+        val = GAPWrap.IntFFE(x)
+        return F(val)
+    end
+
+    # TODO: conversion of elements from extension field; code for this
+    # already exists in `src/Groups/matrices/iso_oscar_gap.jl` but it may require
+    # precomputations, which we should cache
+
+    error("TODO: add support for FFE outside the prime field")
+end
+
+##
+## matrix conversion
+##
+
+function __ensure_gap_matrix(obj::GapObj)
+    GAPWrap.IsMatrixOrMatrixObj(obj) || throw(ArgumentError("<obj> is not a GAP matrix"))
+end
+
 ##
 ## matrix of GAP integers to `fmpz_mat`
 ##
 function fmpz_mat(obj::GapObj)
-    m = Matrix{fmpz}(obj)
-    return matrix(ZZ, m)
+    __ensure_gap_matrix(obj)
+    nrows = GAPWrap.NrRows(obj)
+    ncols = GAPWrap.NrCols(obj)
+    m = zero_matrix(ZZ, nrows, ncols)
+    for i in 1:nrows, j in 1:ncols
+        x = obj[i,j]
+        @inbounds m[i,j] = x isa Int ? x : fmpz(x::GapObj)
+    end
+    return m
 end
 
-GAP.gap_to_julia(::Type{fmpz_mat}, obj::GapObj) = fmpz_mat(obj)
-matrix(R::FlintIntegerRing, obj::GapObj) = fmpz_mat(obj)
+GAP.gap_to_julia(::Type{fmpz_mat}, obj::GapObj) = fmpz_mat(obj) # TODO: deprecate/remove this
 
 ##
 ## matrix of GAP rationals or integers to `fmpq_mat`
 ##
 function fmpq_mat(obj::GapObj)
-    m = Matrix{fmpq}(obj)
-    return matrix(QQ, m)
+    __ensure_gap_matrix(obj)
+    nrows = GAPWrap.NrRows(obj)
+    ncols = GAPWrap.NrCols(obj)
+    m = zero_matrix(QQ, nrows, ncols)
+    for i in 1:nrows, j in 1:ncols
+        x = obj[i,j]
+        @inbounds m[i,j] = x isa Int ? x : fmpq(x::GapObj)
+    end
+    return m
+end
+
+GAP.gap_to_julia(::Type{fmpq_mat}, obj::GapObj) = fmpq_mat(obj) # TODO: deprecate/remove this
+
+##
+## generic matrix() method for GAP matrices which converts each element on its
+## own: this is inefficient but almost always works, so we use it as our base
+## case
+##
+function matrix(R::Ring, obj::GapObj)
+    # TODO: add special code for compressed matrices, resp. MatrixObj, so that
+    # we can perform the characteristic check once, instead of nrows*ncols
+    # times
+    __ensure_gap_matrix(obj)
+    nrows = GAPWrap.NrRows(obj)
+    ncols = GAPWrap.NrCols(obj)
+    m = zero_matrix(R, nrows, ncols)
+    for i in 1:nrows, j in 1:ncols
+        x = obj[i,j]::Union{Int,GapObj,GAP.FFE} # type annotation so Julia generates better code
+        @inbounds m[i,j] = R(x)
```

Comment:
@tthsqe12 the extra alloc comes from this line, where each Int is converted to fmpz before being assigned. This is avoided as a special case in the fmpz_mat constructor above:
```
julia> gapobj = GAP.@gap [[1,1,1,1,1,1,1]]; @btime matrix(ZZ, gapobj)

  1.215 μs (10 allocations: 272 bytes)
[1   1   1   1   1   1   1]

julia> gapobj = GAP.@gap [[1,1,1,1,1,1,1]]; @btime fmpz_mat(gapobj)
  1.376 μs (3 allocations: 160 bytes)
[1   1   1   1   1   1   1]
```
If all our matrix types are guaranteed to provide setindex for Int, then I could add the special case here, too

Reply:
If we do want to make that guarantee, then I guess our conformance tests should cover it 😜

Reply:
oopsie here?
```
julia> gapobj = GAP.@gap -2^100
GAP: -1267650600228229401496703205376

julia> @which fmpz(GAP.ADDR_OBJ(gapobj), div(GAP.SIZE_OBJ(gapobj), sizeof(Int)))
fmpz(x::Ptr{UInt64}, len::Int64) in Nemo at /home/schultz/.julia/dev/Nemo/src/flint/FlintTypes.jl:50

julia> fmpz(gapobj)
1267650600228229401496703205376
```

Reply:
As for the extra allocations, indeed I was calling `matrix(ZZ, ...` instead of `fmpz_mat(` :(. Everything is as expected with three allocations for any matrixs of small integers. :)

Reply:
As for the `mat[i,j] = Int`, I think that is always supposed to work regardless of the presence of a special setindex!, so why not special case it here if boxed `Int`s are coming out of many gap.jl functions.

Reply:
I've applied the following change.
```suggestion
        @inbounds m[i,j] = x isa Int ? x :R(x)
```
Before that:
```
julia> gapobj = GAP.@gap [[1,1,1,1,1,1,1]]; @btime matrix(ZZ, gapobj);
  827.163 ns (10 allocations: 272 bytes)
```
After:
```
julia> gapobj = GAP.@gap [[1,1,1,1,1,1,1]]; @btime matrix(ZZ, gapobj);
  752.154 ns (3 allocations: 160 bytes)
```

Thanks for catching the error converting negative GAP ints, this is quite embarrassing. I'll fix this in a separate PR (and add more tests)